### DISCUSSION TITLE
How can I send a noteOn event with zero velocity?

### DISCUSSION THREAD

**Discussion by happy77** [orchetect/swift-midi#190] [Q&A] [answered]
I have been trying different things for a day now, but I just cannot figure out how to send a noteOn event to my Behringer Xtouch with zero velocity, it keeps changing the velocity to one before sending out.

If did find the hint that midi2 notes converted to midi1 will be transformed changing the velocity but I really don't know how I can avoid this behaviour....

        
             try? midiOutputConnection?.send(
                event: .noteOn(
                    MIDINote(pitch),
                    velocity: .midi1(UInt7(velocity)),
                    channel: UInt4(channel),
                    midi1ZeroVelocityAsNoteOff: false
                )
            )

Any tipps on what I can do to actually send a noteOn with velocity = 0 ?

Thanks, Jan

**Comment by orchetect** [comment, marked as answer]
Yes it appears Core MIDI is doing that translation (which is part of the MDI 2.0 spec).

The only workaround may be to force older non-UMP MIDI 1.0 API.

Just before calling `midiManager.start()`, try this:

```swift
midiManager.preferredAPI = .legacyCoreMIDI
```

**Reply by happy77** [reply]
Wooohoo, works :)

Will this continue to work, or do you think this might become deprecated? As It seems the Lights off function on the controller depends on this?

Thank you so much for your quick reply and fix!

**Reply by orchetect** [reply]
Apple essentially deprecated the legacy Core MIDI API as of macOS 11 Big Sur so it's only a matter of time before it's fully obsoleted and no longer available. MIDIKit will support rolling back to it for as long as it's still possible.

Unfortunately there are a few things about the MIDI 2.0 spec that were designed to be "lossy" and not translate perfectly between 1.0 and 2.0. It's up to Behringer to release a firmware update that can remap its reliance on Note-On velocity 0 to some other MIDI event that's compatible with MIDI 2.0.

Have you tried a Note Off with velocity zero (while using MIDI 2.0)? Perhaps the Behringer will recognize that as the same thing.

**Reply by happy77** [reply]
Hi,

yes I tried noteOff with velocity set to 0, but that did not work. 
I guess  I will need to hope Behringer does release a firmware update before Apple removes legacy support.

Or is it possible to enable sending raw midipakets so kinda working around the spec by passing rawbytes, kinda like 0x90, 0x60, 0x00 ? (As far as I understand that would be noteon to Chan 0 with velocity zero…?)

**Reply by orchetect** [reply]
> working around the spec by passing rawbytes

No, MIDIKit is already encoding messages correctly. The mitigating factor is Core MIDI which we have no control over - it's doing the translation that you are encountering.

I ran some bare-metal tests to confirm the behavior. These results are from sending a Note On event using each API and receiving the sent event using each API in all possible combinations.

Send: Note On `velocity: .midi1(0), midi1ZeroVelocityAsNoteOff: false`

- New API → New API: velocity 0 received
- New API → Old API: velocity 1 received (Core MIDI translating) ←-- 💡 **The problem you had**
- Old API → Old API: velocity 0 received ←-- 💡 **The solution I proposed with using legacy API**
- Old API → New API: velocity 0 received

Send: Note On `velocity: .midi2(0), midi1ZeroVelocityAsNoteOff: false`

- New API → New API: velocity 0 received
- New API → Old API: velocity 1 received (Core MIDI translating)
- Old API → Old API: velocity 1 received (MIDIKit translating `midi2()` value when sending)
- Old API → New API: velocity 1 received (MIDIKit translating `midi2()` value when sending)
